### PR TITLE
Added support for newer docker-compose versions

### PR DESCRIPTION
**Pull request by 3mb3dw0rk5**
Started using your project which is what I was looking for a long time to backup my server configurations.

So I started to align the code with the current docker-compose and docker-api releases. Also moved on to use python 3.7.

Hope these fixes help others as well. I would also contribute to further feature implementations like restore procedure, removing the backup-memory limitation etc.

### CODE REVIEW

**Comment by funkyfuture** [comment]
ciao, thanks for taking care. as you noticed i haven't a while. i hope i can do a review later today.

> I would also contribute to further feature implementations like restore procedure, removing the backup-memory limitation etc.

that'd be definitively welcome. to be sure of expectations, i generally like roadmaps and design discussions. feel free to open an issue in this regard, or by opening a separate PR to add a `ROADMAP.md`.

one thing i will ask in the review is to keep support for Python 3.6. on the other hand, i think support for Docker-Compose < 1.10 can be safely removed. i don't have an overview atm, but my tendency is to remove support for versions where Docker's `copy` endpoint can't be used, because iirc there are some rather hackish parts of the code where `copy` would certainly be preferable. anyway, if you had any insights regarding compatibility and their maintenace cost, i'd be interested. maybe a compatibility policy would be helpful (i usually install Docker-Compose from the PyPI, not the ones that come with distributions).

**Comment by funkyfuture** [comment]
please add an `AUTHORS.md`!

**Comment by 3mb3dw0rk5** [comment]
By the way I changed python compatibility to >= 3.5 which is used on old debian stretch (faced that issue on my pi). But tests seems to run with 3.5, so i dont know if there may be further issues.

**Comment by 3mb3dw0rk5** [comment]
> please add an `AUTHORS.md`!

Any templates you have in mind?

**Comment by funkyfuture** [comment]
> Any templates you have in mind?

```md
# Authors

- Frank Sachsenheim
- …
```

> By the way I changed python compatibility to >= 3.5

i did remove that in c89e32eb00a460fb2b0465f1ff334ce5580ab256, i don't recall why. but it's okay to [keep it for a while](https://devguide.python.org/#status-of-python-branches).

File: compose_dump/backup.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
-from collections import Mapping, OrderedDict, Sequence
+from collections import OrderedDict, Sequence
```
Comment (funkyfuture) [inline compose_dump/backup.py]: shouldn't it be the other way around? see the first Changes note [here](https://docs.python.org/3.8/library/collections.html).

File: compose_dump/backup.py [hunk 2]
```diff
@@ -230,9 +230,12 @@ def store_project_volumes(ctx):
             if container is None:
                 log.critical('Found no container that uses project volume %s' % name)
                 continue
-            response, stat = ctx.project.client.get_archive(container.id, path)
+            bits, stat = ctx.project.client.get_archive(container.id, path)
             archive_name = name + '.tar'
-            ctx.storage.write_file(response.stream, archive_name, namespace='volumes/project')
+            if hasattr(bits, 'stream'):
```
Comment (funkyfuture) [inline compose_dump/backup.py]: please add a note `TODO remove when support for Docker-Compose < 1.19 is dropped`.

File: compose_dump/backup.py [hunk 3]
```diff
@@ -273,8 +276,11 @@ def store_services_volumes(ctx, mounted_paths):
                 continue
             for path in internal_volumes:
                 archive_name = hash_string(service.name.upper() + path) + '.tar'
-                response, stat = ctx.project.client.get_archive(container.id, path)
-                ctx.storage.write_file(response.stream, archive_name, namespace='volumes/services')
+                bits, stat = ctx.project.client.get_archive(container.id, path)
```
Comment (funkyfuture) [inline compose_dump/backup.py]: as above

File: compose_dump/storage.py [hunk 4]
```diff
@@ -124,6 +125,13 @@ def write_file(self, data, dst, namespace='.'):
                 size += len(chunk)
                 buffer.write(chunk)
             buffer.seek(0)
+        elif isinstance(data, types.GeneratorType):
```
Comment (funkyfuture) [inline compose_dump/storage.py]: would the previous branch (`callable(data)`) also be obsolete when `docker-py>=3` is used? if so, a note when to remove it, would be very helpful.

File: setup.py [hunk 5]
```diff
@@ -36,7 +36,7 @@
     author_email='funkyfuture@riseup.net',
     license='ISC',
     platforms=["any"],
-    install_requires=['docker-compose>=1.7,<1.20'],
+    install_requires=['docker-compose>=1.7'],
```
Comment (funkyfuture) [inline setup.py]: please keep the constraint for the latest supported version of Docker-Compose.

File: compose_dump/backup.py (hunk 1 above)
Comment (3mb3dw0rk5) [reply compose_dump/backup.py]: Strangely the warning was not reported anymore. But true, Sequence is also located in collections.abc. OrderedDict stays in collections.

File: compose_dump/storage.py (hunk 4 above)
Comment (3mb3dw0rk5) [reply compose_dump/storage.py]: Yes you are right. `callable(data)` is only needed when `bits.stream` is passed to this function. `stream` is a function allowing to specify a chunk size of the returned data generator. This data generator will be returned directly with docker-py>=3.0.0.

**Comment by 3mb3dw0rk5** [comment]
Committed the changes from your review :+1:

**Comment by funkyfuture** [comment]
thanks!

**Pull request merged**